http-clients/netty-nio-client/src/test/java/software/amazon/awssdk/http/nio/netty/internal/GoAwayTest.java

```diff
@@ -0,0 +1,331 @@
+                frameWriter().writeHeaders(ctx, streamId, new DefaultHttp2Headers().add("content-length", "0").status("204"), 0, true, ctx.newPromise());
+                ctx.flush();
+            }
+        };
+
+        endpointDriver = new SimpleEndpointDriver(frameListenerSupplier);
+        endpointDriver.init();
+
+        netty = NettyNioAsyncHttpClient.builder()
+                .eventLoopGroup(SdkEventLoopGroup.builder().numberOfThreads(1).build())
+                .protocol(Protocol.HTTP2)
+                .build();
+
+        sendGetRequest().join();
+
+        // Note: It's possible the initial request can cause the client to allocate more than 1 channel
+        int initialChannelNum = endpointDriver.channels.size();
+
+        // Send GOAWAY to all the currently open channels
+        endpointDriver.channels.forEach(ch -> endpointDriver.goAway(ch, 1));
+
+        // Need to give a chance for the streams to get closed
+        Thread.sleep(1000);
+
+        // Since the existing channels are now invalid, this request should cause a new channel to be opened
+        sendGetRequest().join();
+
+        assertThat(endpointDriver.channels).hasSize(initialChannelNum + 1);
+    }
+
+    // The client should not close streams that are less than the 'last stream
+    // ID' given in the GOAWAY frame since it means they were processed fully
+    @Test
+    public void execute_goAwayReceived_lastStreamId_lowerStreamsNotClosed() throws InterruptedException {
+        ConcurrentHashMap<String, Set<Integer>> channelToStreams = new ConcurrentHashMap<>();
+
+        CompletableFuture<Void> stream3Received = new CompletableFuture<>();
+        CountDownLatch allRequestsReceived = new CountDownLatch(2);
+        byte[] getPayload = "go away!".getBytes(StandardCharsets.UTF_8);
+        Supplier<Http2FrameListener> frameListenerSupplier = () -> new TestFrameListener() {
+            @Override
+            public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers headers, int streamDependency, short weight, boolean exclusive, int padding, boolean endStream) {
+                channelToStreams.computeIfAbsent(ctx.channel().id().asShortText(), (k) -> Collections.newSetFromMap(new ConcurrentHashMap<>())).add(streamId);
+
+                if (streamId == 3) {
+                    stream3Received.complete(null);
+                }
+
+                if (streamId < 5) {
+                    Http2Headers outboundHeaders = new DefaultHttp2Headers()
+                            .status("200")
+                            .add("content-type", "text/plain")
+                            .addInt("content-length", getPayload.length);
+
+                    frameWriter().writeHeaders(ctx, streamId, outboundHeaders, 0, false, ctx.newPromise());
+                    ctx.flush();
+                }
```

Comment:
I'm probably missing something. Why 3 and 5 when there's only 2 requests?

Reply:
[The client always initiates streams using odd numbers](https://httpwg.org/specs/rfc7540.html#StreamIdentifiers). The first stream used to establish the H2 stream is 1, so the next two are 3 and 5.

Reply:
Ah, cool!